Article for day 14. (#73)
scalacenter/scala-advent-of-code: docs/puzzles/day14.md

For any string longer than 2 chars, we have the following property:

$$
S(x_1 x_2 x_3 \cdots x_p, n) = \Sigma_{i=1}^{p-1} (S(x_i x_{i+1}, n))
$$

In other words, $S$ for a string is the *sum* (a multiset sum) of $S$ for all the adjacent pairs in the string, with the same number of iterations $n$.
That is because each initial pair expands independently of the others.
Each initial char is counted exactly once in the final frequencies because it is counted as part of the expansion of the pair on its left, and not the expansion of the pair on its right (we always exclude the first char).

As a particular case of the above, for a string of 3 chars $xzy$, we have

$$
S(xzy, n) = S(xz, n) + S(zy, n)
$$

For strings of length 2, we have two cases: $n = 0$ and $n > 0$.

Base case: a pair $xy$, and $n = 0$

$$
S(xy, 0) = \{ y \rightarrow 1 \} \text{ for all } x, y \text{ (by definition)}
$$

Inductive case: a pair $xy$, and $n > 0$

$$
\begin{aligned}
S(xy, n) & = S(xzy, n-1) \text{ where $z$ is the insertion char for the pair $xy$ (by definition)} \\
         & = S(xz, n-1) + S(zy, n-1) \text{ -- the particular case of 3 chars above}
\end{aligned}
$$

> 

Reviewer: Could we add something like:
`This means that occurrences of letters after xy has produced it's final polimer are equal to the occurences of xzy after 1 less step`

> Now that we have an inductive defition of $S(xy, n)$ for all pairs, we can write an interative algorithm that computes that for all possible pairs $xy$, for $n \in \lbrack 0, 40 \rbrack$ :

Reviewer 2: ```suggestion
Now that we have an inductive definition of $S(xy, n)$ for all pairs, we can write an iterative algorithm that computes that for all possible pairs $xy$, for $n \in \lbrack 0, 40 \rbrack$ :
```

  val sections = input.split("\n\n")
  val initialPolymer = sections(0).toList
  val insertionRules = sections(1).linesIterator.map(parseRule).toMap
  (initialPolymer, insertionRules)

def parseRule(line: String): (CharPair, Char) =
  line match
    case s"$pairStr -> $inserStr" => (pairStr(0), pairStr(1)) -> inserStr(0)
    case _                        => throw new Exception(s"Cannot parse '$line' as an insertion rule")

def applyRules(polymer: Polymer, rules: InsertionRules): Polymer =
  val pairs = polymer.zip(polymer.tail)
  val insertionsAndSeconds: List[List[Char]] =
    for pair @ (first, second) <- pairs
      yield rules(pair) :: second :: Nil
  polymer.head :: insertionsAndSeconds.flatten
```

<Solver puzzle="day14-part1"/>

## Part 2

The statement of part 2 says to do the same thing with 40 iterations.
Of course, it is not as simple as reusing the first implementation above.
The final polymer's length would exceed 4G of characters, let alone the additional memory of the `List` and the processing time required to build it.
We need to compute the eventual frequencies *without* computing the `finalPolymer` at all.

> Out of the blue, we define the abstract function $S(ab \cdots f, n)$ as being the frequencies of all letters, *except the first one*, in the expansion of $ab \cdots f$ after $n$ steps.

Author: That's what I intended. It's hard to explain how we come up with that function and that definition. It's intuition, that is only rationalized after the fact by the fact that the "proof" works out.

> Out of the blue, we define the abstract function $S(ab \cdots f, n)$ as being the frequencies of all letters, *except the first one*, in the expansion of $ab \cdots f$ after $n$ steps.

Reviewer: So maybe:
```suggestion
Intuitively, let's start with defining the abstract function $S(ab \cdots f, n)$ as being the frequencies of all letters, *except the first one*, in the expansion of $ab \cdots f$ after $n$ steps.
```

?

`out of the blue` doesn't sound right in this situation.

> Out of the blue, we define the abstract function $S(ab \cdots f, n)$ as being the frequencies of all letters, *except the first one*, in the expansion of $ab \cdots f$ after $n$ steps.

Author: OK, I tried a definition formulation:
```md
Let us start by defining the abstract function $S(ab \cdots f, n)$ as being the frequencies of all letters, *except the first one*, in the expansion of $ab \cdots f$ after $n$ steps.
We come up with this function definition based on mathematical intuition.
We can only justify that choice by showing hereafter that it allows us to solve our problem.
```

> Out of the blue, we define the abstract function $S(ab \cdots f, n)$ as being the frequencies of all letters, *except the first one*, in the expansion of $ab \cdots f$ after $n$ steps.

Reviewer: Thanks!

With the example insertion rules, and for the example string `NNCB`, we would have:

$$
\begin{aligned}
S(\text{NNCB}, 0) & = \{ \text{B} \rightarrow 1, \text{C} \rightarrow 1, \text{N} \rightarrow 1 \} \\
S(\text{NNCB}, 1) & = S(\text{NCNBCHB}, 0) \\
                  & = \{ \text{B} \rightarrow 2, \text{C} \rightarrow 2, \text{H} \rightarrow 1, \text{N} \rightarrow 1 \}
\end{aligned}
$$

> Note that one `N` is excluded from the frequencies every time---the one at the beginning of the string.

Reviewer 2: It’s supposed to be a dash, but it’s not correctly rendered:
![Screenshot from 2021-12-14 18-05-47](https://user-images.githubusercontent.com/332812/146045613-91215fe5-3cd2-4f8c-9904-b0ae0ec03743.png)

> Note that one `N` is excluded from the frequencies every time---the one at the beginning of the string.

Reviewer 2: You may want to use parens instead.
```suggestion
Note that one `N` is excluded from the frequencies every time (ie, the one at the beginning of the string).
```

> Note that one `N` is excluded from the frequencies every time---the one at the beginning of the string.

Author: I used parens ;)
